File: Leetcode/P009_Palindrome_Number.cpp
/*
 * https://leetcode.com/problems/palindrome-number/description/
 * 给定一个整数, 判断它是不是回文数
 * 先转为字符串, 然后对其reverse, 最后判断内容是否相同
*/

// #include <iostream>
// #include <string>
// #include <sstream>
// #include <algorithm>
// using namespace std;

class Solution {
public:
	string num2str(int x){
		stringstream ss;
		ss<<x;
		string s;
		ss>>s;
		return s;
	}

	bool isPalindrome(int x) {
		string s, s1;
		s = s1 = num2str(x);
		reverse(s.begin(), s.end());
		return s == s1;
	}
};


// int main(){
// 	ios::sync_with_stdio(false);
// 	cin.tie(0);

// 	Solution s;
// 	cout<<s.isPalindrome(1)<<'\n';
// 	cout<<s.isPalindrome(121)<<'\n';
// 	cout<<s.isPalindrome(2147483647)<<'\n';

// 	return 0;
// }

File: LeetCode/001-100/P009_Palindrome_Number.java
/*
https://leetcode.com/problems/palindrome-number/description/
给定一个整数, 判断它是不是回文数
先转为字符串, 然后对其reverse, 最后判断内容是否相同

输入参数 x 的范围
-2^31 <= x <= 2^31 - 1
如果使用 reverse 数字的方法, 可能会出现溢出 (参考 P007 ), 需要用 long 来存 reverse 之后的数字

*/

class Solution {
    public boolean isPalindrome(int x) {
        if ( x < 0 ) {
            return false;
        }
        long a = x;
        long b = 0;
        while ( x > 0 ) {
            b = b * 10 + x % 10;
            x /= 10;
        }
        return a == b;
    }

    public boolean isPalindrome2(int x) {
        if ( x < 0 ) {
            return false;
        }
        String s = Integer.toString(x);
        int left = 0, right = s.length() - 1;
        while ( left < right ) {
            if ( s.charAt(left) != s.charAt(right) ) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    public boolean isPalindrome1(int x) {
        if ( x < 0 ) {
            return false;
        }
        String s1 = Integer.toString(x);
        String s2 = new StringBuilder(s1).reverse().toString();
        return s1.equals(s2);
    }
}